File: backend/src/routes/budget.rs
use actix_web::{delete, get, patch, post, web, HttpResponse};
use sqlx::PgPool;

use crate::{
    auth::Claims,
    error::Result,
    models::budget::{Monthly, MonthlyAdd, MonthlyDelete, MonthlyUpdate},
    repo::budget::BudgetRepo,
};

#[get("/monthly")]
pub async fn get_all(pool: web::Data<PgPool>, claims: Claims) -> Result<web::Json<Vec<Monthly>>> {
    let mut connection = pool.acquire().await?;

    BudgetRepo::monthly_get_all(connection.as_mut(), claims)
        .await
        .map(web::Json)
}

#[post("/monthly")]
pub async fn add(
    pool: web::Data<PgPool>,
    claims: Claims,
    monthly: web::Json<MonthlyAdd>,
) -> Result<HttpResponse> {
    let mut connection = pool.acquire().await?;

    BudgetRepo::insert_monthly(connection.as_mut(), claims, monthly.into_inner()).await?;

    Ok(HttpResponse::Ok().finish())
}

#[patch("/monthly")]
pub async fn update(
    pool: web::Data<PgPool>,
    claims: Claims,
    monthly: web::Json<MonthlyUpdate>,
) -> Result<HttpResponse> {
    let mut connection = pool.acquire().await?;

    BudgetRepo::update_monthly(connection.as_mut(), claims, monthly.into_inner()).await?;

    Ok(HttpResponse::Ok().finish())
}

#[delete("/monthly")]
pub async fn delete_monthly(
    pool: web::Data<PgPool>,
    claims: Claims,
    id: web::Json<MonthlyDelete>,
) -> Result<HttpResponse> {
    let mut connection = pool.acquire().await?;

    BudgetRepo::delete_monthly(connection.as_mut(), claims, &id.into_inner().delete).await?;

    Ok(HttpResponse::Ok().finish())
}

File: frontend/src/api/Budget.ts
import axios, { AxiosResponse } from "axios";

export type Monthly = {
    id: string;
    position: string;
    debit: number;
    credit: number;
}

export type AddMonthly = {
    position: string;
    debit?: number;
    credit?: number;
}

export type UpdateMonthly = {
    id: string;
    position?: string;
    debit?: number;
    credit?: number;
}

export function getMonthly(): Promise<AxiosResponse<Monthly[]>> {
    return axios.get("/api/budget/monthly");
}

export function addMonthly(body: AddMonthly) {
    return axios.post("/api/budget/monthly", body);
}

export function updateMonthly(body: UpdateMonthly) {
    return axios.patch("/api/budget/monthly", body);
}

export function deleteMonthly(ids: string[]) {
    return axios.delete(`/api/budget/monthly`, { data: { delete: ids } });
}